README.md
# django-luoji-accounts


Overview
--------
django-luoji-accounts adds a number of indispensible financial account-related capabilities to your Django application, including:

* Account

* Sub Account

* AccountType

* SystemCode

* DeviceType

* Transfer

* Transaction 

Please note, this app was developed under Django==1.8.2.

Quick Start
-----------
django-luoji-accounts contains a low level account and sub accounts system that you can get up and running right away. 

To install Django-Accounts from PyPI::
```bash
pip install django-luoji-accounts
```
Alternatively, you can also install Django-luoji-Accounts from GitHub::
https://github.com/hereischen/django-luoji-accounts

Add 'accounts' to your INSTALLED_APPS in your settings.py::

```python
INSTALLED_APPS = (
...
'accounts',
...
)
```
Include the django-luoji-accounts URLs in your urls.py::

```python
urlpatterns = patterns(
...
(r'^accounts/', include('accounts.urls')),
...
)
```

To customise for your specific usage, create a wrapper in your project, for example, project/account_wrapper.py, it would be:

```python
from accounts.models import GeneralAccountManager

class AudioAccountManager(GeneralAccountManager):

    """
    音频账户管理类
    """
    def _get_or_create_sub_account_for_jcc(self, user_id, account_type, device_type):
        """
        查询或创建
        """
        return self.get_or_create_sub_account(user_id=user_id,
                                              account_type=account_type,
                                              sys_code='AUDIO',
                                              device_type=device_type,
                                              currency='JCC'
                                              )

    def _get_sub_account_for_jcc(self, user_id, account_type, device_type):
        """
        查询
        """
        return self.get_sub_account(user_id=user_id,
                                    account_type=account_type,
                                    sys_code='AUDIO',
                                    device_type=device_type,
                                    currency='JCC'
                                    )

    def get_or_create_buyer_account_for_android(self, user_id):
        return self._get_or_create_sub_account_for_jcc(user_id,'BUYER', 'ANDROID')

    def get_or_create_cost_account_for_android(self):
        return self._get_or_create_sub_account_for_jcc('9999', 'COST', 'ANDROID')


    def get_buyer_account_for_android(self, user_id):
        return self._get_sub_account_for_jcc(user_id, 'BUYER', 'ANDROID')

    def get_cost_account_for_android(self):
        return self._get_sub_account_for_jcc('9999', 'COST', 'ANDROID')

```

To call the methods :

```python
# import required models e.g.. Account, SubAccount

# import your wrapper e.g. from project/account_wrapper.py
from project.account_wrapper import AudioAccountManager

# query or create an android buyer account 查询或创建安卓买家
# return one if it exists
# create and return one if it does not exist
ba = AudioAccountManager().get_or_create_buyer_account_for_android(
    '100010')

# query an android buyer account 查询安卓买家
# return one if it exists
# arise exceptions if one does not exist
ba2 = AudioAccountManager().get_buyer_account_for_android(
    '100010')

# query or create an android cost account  查询或创建安卓赠送账户
ga = AudioAccountManager().get_or_create_cost_account_for_android()


# query an android cost account 查询安卓赠送账户
ga2 = AudioAccountManager().get_cost_account_for_android()


```


That's it!

CHANGES.rst
Changelog
==============================
0.1.0 - Aug.07, 2015
------------------------------
- New initial mirgations

0.0.9 - Aug.07, 2015
------------------------------
- added GenericForeignKey and ContentType in TransferManager

0.0.8 - Aug.07, 2015
------------------------------
- Changed AUDIO to IGET in fixtures

0.0.7 - Jul.30, 2015
------------------------------
- added fixtures package
- added migrations package

0.0.6 - Jul.30, 2015
------------------------------
- bugs fixing

0.0.5 - Jul.30, 2015
------------------------------
- added new field 'user_id' for Account and SunAccount
- changed _validate_user_id method in GeneralAccountManager
- added _validate_merchant_account_user_id in GeneralAccountManager


0.0.4 - Jul.29, 2015
------------------------------
- _validate_user_id method added in GeneralAccountManager


0.0.3 - Jul.29, 2015
------------------------------

- bug fixing

0.0.2 - Jul.29, 2015
------------------------------

- chanegd readme


0.0.1 - Jul.29, 2015
------------------------------

- Initail commit



accounts/tests.py
from django.test import TestCase
from accounts.models import Account

from decimal import Decimal as D
from jccoin.account_wrapper import AudioAccountManager


# Create your tests here.
class AccountsTestCase(TestCase):
    fixtures = ['accounts.json']

    def test_get_or_create_buyer_account_for_audio_android(self):
        ba = AudioAccountManager().get_or_create_buyer_account_for_android(
            '100010')
        self.assertEquals(ba.account_number, '100401100010')

    def test_get_or_create_buyer_account_for_audio_ios(self):
        bi = AudioAccountManager().get_or_create_buyer_account_for_ios(
            '100010')
        self.assertEquals(bi.account_number, '100402100010')

    def test_get_or_create_seller_account_for_audio_android(self):
        sa = AudioAccountManager().get_or_create_seller_account_for_android(
            '100010')
        self.assertEquals(sa.account_number, '110401100010')

    def test_get_or_create_seller_account_for_audio_ios(self):
        si = AudioAccountManager().get_or_create_seller_account_for_ios(
            '100010')
        self.assertEquals(si.account_number, '110402100010')

    def test_sub_account_quantity(self):

        AudioAccountManager().get_or_create_buyer_account_for_android(
            '100010')
        AudioAccountManager().get_or_create_buyer_account_for_ios(
            '100010')
        AudioAccountManager().get_or_create_seller_account_for_android(
            '100010')
        AudioAccountManager().get_or_create_seller_account_for_ios(
            '100010')

        b_quantity = Account.objects.get(
            account_number='10100010').sub_account_quantity
        a_quantity = Account.objects.get(
            account_number='11100010').sub_account_quantity

        self.assertEquals(2, b_quantity)
        self.assertEquals(2, a_quantity)

    def test_get_or_create_intermediary_account_for_audio(self):
        ia = AudioAccountManager(
        ).get_or_create_intermediary_account_for_android()
        ii = AudioAccountManager(
        ).get_or_create_intermediary_account_for_ios()
        self.assertEquals(ia.account_number, '5104019999')
        self.assertEquals(ii.account_number, '5104029999')

        quantity = Account.objects.get(
            account_number='519999').sub_account_quantity

        self.assertEquals(quantity, 2)

    def test_get_or_create_cost_account_for_audio(self):
        ia = AudioAccountManager(
        ).get_or_create_cost_account_for_android()
        ii = AudioAccountManager(
        ).get_or_create_cost_account_for_ios()
        self.assertEquals(ia.account_number, '5204019999')
        self.assertEquals(ii.account_number, '5204029999')

        quantity = Account.objects.get(
            account_number='529999').sub_account_quantity

        self.assertEquals(quantity, 2)

    def test_get_or_create_guarantee_account_for_audio(self):
        ia = AudioAccountManager(
        ).get_or_create_guarantee_account_for_android()
        ii = AudioAccountManager(
        ).get_or_create_guarantee_account_for_ios()
        self.assertEquals(ia.account_number, '5304019999')
        self.assertEquals(ii.account_number, '5304029999')

        quantity = Account.objects.get(
            account_number='539999').sub_account_quantity

        self.assertEquals(quantity, 2)

    def test_get_or_create_income_account_for_audio(self):
        ia = AudioAccountManager(
        ).get_or_create_income_account_for_android()
        ii = AudioAccountManager(
        ).get_or_create_income_account_for_ios()
        self.assertEquals(ia.account_number, '5404019999')
        self.assertEquals(ii.account_number, '5404029999')

        quantity = Account.objects.get(
            account_number='549999').sub_account_quantity

        self.assertEquals(quantity, 2)

    def test_get_or_create_fronzenfund_account_for_audio(self):
        ia = AudioAccountManager(
        ).get_or_create_fronzenfund_account_for_android()
        ii = AudioAccountManager(
        ).get_or_create_fronzenfund_account_for_ios()
        self.assertEquals(ia.account_number, '5504019999')
        self.assertEquals(ii.account_number, '5504029999')

        quantity = Account.objects.get(
            account_number='559999').sub_account_quantity

        self.assertEquals(quantity, 2)

    def test_get_or_create_gift_account_for_audio(self):
        ia = AudioAccountManager(
        ).get_or_create_gift_account_for_android()
        ii = AudioAccountManager(
        ).get_or_create_gift_account_for_ios()
        self.assertEquals(ia.account_number, '5604019999')
        self.assertEquals(ii.account_number, '5604029999')

        quantity = Account.objects.get(
            account_number='569999').sub_account_quantity

        self.assertEquals(quantity, 2)

    # def test_can_not_get_buyer_account_for_android(self):
    #     a = AudioAccountManager().get_buyer_account_for_android('100010')
    #     self.assertIsInstance(a, Exception)

    def test_can_get_buyer_account_for_android(self):
        AudioAccountManager().get_or_create_buyer_account_for_android(
            '100010')
        ba = AudioAccountManager().get_buyer_account_for_android(
            '100010')

        self.assertEquals(ba.account_number, '100401100010')
        self.assertEquals(ba.balance, D('0.000000'))


accounts/models.py
# -*- coding: utf-8 -*-
import logging
from decimal import Decimal as D

from django.db import models
from django.db.models import Q
from django.db.models import Sum
from django.utils.translation import ugettext_lazy as _
from django.contrib.contenttypes.fields import GenericForeignKey
from django.contrib.contenttypes.models import ContentType

logger = logging.getLogger(__name__)

ACCOUNT_STATUS = (
    ('OPEN', '开启'),
    ('FROZEN', '冻结'),
    ('CLOSED', '关闭'),
)

CURRENCY = (
    ('CNY', '人民币'),
    ('JCC', '节操币'),
)


class GeneralAccountManager(object):

    """
    项目账户管理通用类
    """

    def _format_user_id(self, user_id):
        # 用户id都是数字
        if user_id.isdigit():
            # 不能超过9位,不足9位,则首位补9,之后补0＋user_id 直到9位
            if len(user_id) == 9:
                return user_id
            elif len(user_id) < 9:
                user_id = '9' + '{0}'.format(user_id.zfill(8))
                return user_id
            else:
                raise ValueError(
                    "Invalid user id [%s]. User id must not exceed 9 digits." % user_id)
        else:
            raise ValueError(
                "Invalid user id [%s]. User id must be all digits." % user_id)

    def _generate_account_number(self, user_id, account_type,
                                 sys_code, device_type):
        """
        生成account_number:
        总账户 账户号 = 两位账户类型编码  + rf_user id
        子账户 账户号 = 两位账户类型编码 + 两位系统(项目)类型编码 + 两位设备类型编码 + rf_user id
        """
        # 账户类型编码
        account_code = AccountType.objects.get(
            account_type_en=account_type).mapping_code
        # 系统(项目)类型编码
        system_code = SystemCode.objects.get(
            sys_code_en=sys_code).mapping_code
        # 设备类型编码
        device_code = DeviceType.objects.get(
            device_type_en=device_type).mapping_code

        rf_user_id = self._format_user_id(user_id)

        # 总账户 账户号
        account_number = account_code + rf_user_id

        # 子账户 账户号
        sub_account_number = account_code + \
            system_code + device_code + rf_user_id

        return account_number, sub_account_number

    def _is_account_exist(self, account_number):
        return Account.objects.filter(account_number=account_number)

    def _create_account(self, user_id, account_type,
                        sys_code, device_type):

        account_number = self._generate_account_number(user_id=user_id,
                                                       account_type=account_type,
                                                       sys_code=sys_code,
                                                       device_type=device_type)[0]

        account = Account.objects.create(account_number=account_number,
                                         user_id=user_id,
                                         sub_account_quantity=1,
                                         )
        account.save()
        logger.info('Create account >> %s.' % account)

        return account

    def get_or_create_sub_account(self, user_id, account_type,
                                  sys_code, device_type, currency):

        account_number, sub_account_number = self._generate_account_number(user_id=user_id,
                                                                           account_type=account_type,
                                                                           sys_code=sys_code,
                                                                           device_type=device_type)

        if not self._is_account_exist(account_number):
            # 创建一个新的总账户,并使sub_account_quantity为1
            account = self._create_account(user_id=user_id, account_type=account_type,
                                           sys_code=sys_code, device_type=device_type)

            # 子账户
            sub_account, flg = SubAccount.objects.get_or_create(account_number=sub_account_number,
                                                                user_id=user_id,
                                                                currency=currency,
                                                                account=account
                                                                )

            if flg:
                # sub_account.save()
                logger.info('Create sub account >> %s.' % sub_account)
            else:
                logger.info('Get sub account >> %s.' % sub_account)

            return sub_account

        else:
            account = self._is_account_exist(account_number)[0]

            # 子账户
            sub_account, flg = SubAccount.objects.get_or_create(account_number=sub_account_number,
                                                                user_id=user_id,
                                                                currency=currency,
                                                                account=account
                                                                )

            # 创建子账户时,总账户的子账户数+1
            if flg:
                # sub_account.save()
                logger.info('Create sub account >> %s.' % sub_account)
                account.sub_account_quantity += 1
                account.save()
                logger.info('Update account >> %s.' % account)
            else:
                logger.info('Get sub account >> %s.' % sub_account)

            return sub_account

    def get_sub_account(self, user_id, account_type,
                        sys_code, device_type, currency):

        _, sub_account_number = self._generate_account_number(user_id=user_id,
                                                              account_type=account_type,
                                                              sys_code=sys_code,
                                                              device_type=device_type)

        return SubAccount.objects.get(account_number=sub_account_number)

    def _validate_merchant_account_user_id(self, user_id):
        if user_id == '999999999':
            raise ValueError(
                "Invalid user id [%s] for merchant users. [99999999] is reserved " % user_id)


class Account(models.Model):
    # 注意账户统一保留小数点后六位
    # 账户编号
    account_number = models.CharField(
        max_length=128, unique=True, help_text=u'总账户编号')

    user_id = models.CharField(max_length=128, help_text=u'用户id')

    # 账户创建时间
    time_created = models.DateTimeField(auto_now_add=True, help_text=u'账户创建时间')
    # 账户状态
    status = models.CharField(
        max_length=32, choices=ACCOUNT_STATUS, default='OPEN', help_text=u'账户状态')

    # 自账户数
    sub_account_quantity = models.IntegerField(help_text=u'子账户数目')

    class Meta:
        verbose_name = _('account')
        verbose_name_plural = _('accounts')

    def __unicode__(self):
        return u'Account number: [%s] & sub account quantity is [%d]' % (self.account_number,
                                                                         self.sub_account_quantity)

    def is_open(self):
        return self.status == 'OPEN'

    def is_closed(self):
        return self.status == 'CLOSED'

    def is_frozen(self):
        return self.status == 'FROZEN'


class SubAccount(models.Model):
    account_number = models.CharField(
        max_length=169, unique=True, help_text=u'子账户编号')

    user_id = models.CharField(max_length=128, help_text=u'用户id')

    # 账户创建时间
    time_created = models.DateTimeField(auto_now_add=True, help_text=u'账户创建时间')

    # 货币种类
    currency = models.CharField(
        max_length=8, choices=CURRENCY, default='CNY', help_text=u'交易币种')

    # 余额
    balance = models.DecimalField(decimal_places=6, max_digits=12,
                                  default=D('0.000000'), null=True, help_text=u'账户余额')

    # 账户状态
    status = models.CharField(
        max_length=32, choices=ACCOUNT_STATUS, default='OPEN', help_text=u'账户状态')

    account = models.ForeignKey('accounts.Account',
                                related_name='sub_accounts')

    class Meta:
        verbose_name = _('sub account')
        verbose_name_plural = _('sub accounts')

    def __unicode__(self):
        return u'Sub account number: [%s], balance: %s.' % (self.account_number,
                                                            self.balance)

    def is_open(self):
        return self.status == 'OPEN'

    def is_closed(self):
        return self.status == 'CLOSED'

    def is_frozen(self):
        return self.status == 'FROZEN'

    def is_debit_permitted(self, amount):
        """
        Test if the a debit for the passed amount is permitted
        """
        if self.account_number[:2] in ('10', '11',):
            return self.balance >= amount
        return True

    def cal_source_balance(self, amount):
        self.balance -= amount
        return self.balance

    def cal_destination_balance(self, amount):
        self.balance += amount
        return self.balance

    def num_transactions(self):
        return u'num_transactions : [%s]' % self.transactions.all().count()

    def _balance(self):
        aggregates = self.transactions.aggregate(sum=Sum('amount'))
        sum = aggregates['sum']
        return D('0.00') if sum is None else sum

    @property
    def amount_available(self):
        return self.balance

# =====================fixtures are provided in accounts.json===============


class AccountType(models.Model):

    account_type_en = models.CharField(max_length=128, help_text=u'英文账户类型')
    account_type_cn = models.CharField(max_length=128, help_text=u'中文账户类型')
    mapping_code = models.CharField(
        max_length=32, unique=True, help_text=u'映射编号')

    class Meta:
        verbose_name = _('account type')
        verbose_name_plural = _('account types')

    def __unicode__(self):
        return u'Account type :%s, mapping code:[%s]' % (self.account_type_en, self.mapping_code)


class SystemCode(models.Model):

    sys_code_en = models.CharField(max_length=128, help_text=u'英文系统代码')
    sys_code_cn = models.CharField(max_length=128, help_text=u'中文系统代码')
    mapping_code = models.CharField(
        max_length=32, unique=True, help_text=u'映射编号')

    class Meta:
        verbose_name = _('system code')
        verbose_name_plural = _('system codes')

    def __unicode__(self):
        return u'System code:%s, mapping code:[%s]' % (self.sys_code_en, self.mapping_code)


class DeviceType(models.Model):

    device_type_en = models.CharField(max_length=128, help_text=u'英文设备类型')
    device_type_cn = models.CharField(max_length=128, help_text=u'中文设备类型')
    mapping_code = models.CharField(
        max_length=32, unique=True, help_text=u'映射编号')

    class Meta:
        verbose_name = _('device type')
        verbose_name_plural = _('device types')

    def __unicode__(self):
        return u'Device type: %s, mapping code:[%s]' % (self.device_type_en, self.mapping_code)

# =========================================================================


class TransferManager(models.Manager):

    def create_transfer(self, source, destination, amount, content_object, description=None):
        self.verify_transfer(source, destination, amount)

        logger.debug('From account [%s : %s] to account [%s : %s], transfer amount: [%s] %s.'
                     % (source.account_number, source.balance,
                        destination.account_number, destination.balance,
                        amount, description))

        transfer = self.get_queryset().create(
            source=source,
            destination=destination,
            amount=amount,
            description=description,
            content_object=content_object,
            source_balance=D('0.000000'),
            destination_balance=D('0.000000'),
        )
        transfer.transactions.create(sub_account=source, amount=-amount)
        transfer.transactions.create(sub_account=destination, amount=amount)
        source.cal_source_balance(amount)
        destination.cal_destination_balance(amount)
        source.save()
        destination.save()
        # 存储transfer后的余额
        transfer.source_balance = source.balance
        transfer.destination_balance = destination.balance
        transfer.save()

    def verify_transfer(self, source, destination, amount):
        """
        Test whether the proposed transaction is permitted.  Raise an exception
        if not.
        """
        if amount < 0:
            raise ValueError(u'Debits cannot use a negative amount')
        if not source.is_open():
            raise Exception(u'Source account stauts is not open.')
        if not destination.is_open():
            raise Exception(u'Destination account stauts is not open.')
        if source.account_number == destination.account_number:
            raise Exception(
                u"Source account and destination account cannot be same.")
        if not source.is_debit_permitted(amount):
            raise ValueError(
                u'Account: %s has insufficient Balance.' % source.account_number)

    def get_transfer_details_by_account(self, account):
        return self.get_queryset().filter(Q(source=account) | Q(destination=account)).order_by('-time_created')


class Transfer(models.Model):

    """
    A transfer of funds between two accounts.
    """
    # 转账资金来源方
    source = models.ForeignKey('SubAccount',
                               related_name='source_transfers')
    # 转账资金接受方
    destination = models.ForeignKey('SubAccount',
                                    related_name='destination_transfers')
    # 转账资金来源方的余额
    source_balance = models.DecimalField(decimal_places=6, max_digits=12,
                                         default=D('0.000000'), help_text=u'转账后的来源方余额')
    # 转账资金接收方的余额
    destination_balance = models.DecimalField(decimal_places=6, max_digits=12,
                                              default=D('0.000000'), help_text=u'转账后的接受方余额')

    # 转账金额
    amount = models.DecimalField(decimal_places=6, max_digits=12)

    # 转账描述
    description = models.CharField(max_length=256, null=True)

    # 转账创建时间
    time_created = models.DateTimeField(auto_now_add=True)

    # 建立与withdrawal和trade之间的关系
    content_type = models.ForeignKey(ContentType)
    object_id = models.PositiveIntegerField()
    content_object = GenericForeignKey('content_type', 'object_id')

    objects = TransferManager()

    class Meta:
        verbose_name = _('transfer')
        verbose_name_plural = _('transfers')

    def __unicode__(self):
        return u"Transfer source: %s, destination: %s" % (self.source, self.destination)

    def delete(self, *args, **kwargs):
        raise RuntimeError("Transfers cannot be deleted")


class Transaction(models.Model):
    # Every transfer of money should create two rows in this table.
    # (1) the debit from the source account
    # (2) the credit to the destination account
    transfer = models.ForeignKey('accounts.Transfer',
                                 related_name='transactions')
    sub_account = models.ForeignKey('accounts.SubAccount',
                                    related_name='transactions')

    # The sum of this field over the whole table should always be 0.
    # Credits should be positive while debits should be negative
    amount = models.DecimalField(decimal_places=6, max_digits=12)
    time_created = models.DateTimeField(auto_now_add=True)

    def __unicode__(self):
        return u"Transaction amount: %.6f" % (self.amount)

    class Meta:
        unique_together = ('transfer', 'sub_account')
        verbose_name = _('transaction')
        verbose_name_plural = _('transactions')

    def delete(self, *args, **kwargs):
        raise RuntimeError("Transactions cannot be deleted")


accounts/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import models, migrations
from decimal import Decimal


class Migration(migrations.Migration):

    dependencies = [
        ('contenttypes', '0002_remove_content_type_name'),
    ]

    operations = [
        migrations.CreateModel(
            name='Account',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('account_number', models.CharField(help_text='\u603b\u8d26\u6237\u7f16\u53f7', unique=True, max_length=128)),
                ('user_id', models.CharField(help_text='\u7528\u6237id', max_length=128)),
                ('time_created', models.DateTimeField(help_text='\u8d26\u6237\u521b\u5efa\u65f6\u95f4', auto_now_add=True)),
                ('status', models.CharField(default=b'OPEN', help_text='\u8d26\u6237\u72b6\u6001', max_length=32, choices=[(b'OPEN', b'\xe5\xbc\x80\xe5\x90\xaf'), (b'FROZEN', b'\xe5\x86\xbb\xe7\xbb\x93'), (b'CLOSED', b'\xe5\x85\xb3\xe9\x97\xad')])),
                ('sub_account_quantity', models.IntegerField(help_text='\u5b50\u8d26\u6237\u6570\u76ee')),
            ],
            options={
                'verbose_name': 'account',
                'verbose_name_plural': 'accounts',
            },
        ),
        migrations.CreateModel(
            name='AccountType',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('account_type_en', models.CharField(help_text='\u82f1\u6587\u8d26\u6237\u7c7b\u578b', max_length=128)),
                ('account_type_cn', models.CharField(help_text='\u4e2d\u6587\u8d26\u6237\u7c7b\u578b', max_length=128)),
                ('mapping_code', models.CharField(help_text='\u6620\u5c04\u7f16\u53f7', unique=True, max_length=32)),
            ],
            options={
                'verbose_name': 'account type',
                'verbose_name_plural': 'account types',
            },
        ),
        migrations.CreateModel(
            name='DeviceType',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('device_type_en', models.CharField(help_text='\u82f1\u6587\u8bbe\u5907\u7c7b\u578b', max_length=128)),
                ('device_type_cn', models.CharField(help_text='\u4e2d\u6587\u8bbe\u5907\u7c7b\u578b', max_length=128)),
                ('mapping_code', models.CharField(help_text='\u6620\u5c04\u7f16\u53f7', unique=True, max_length=32)),
            ],
            options={
                'verbose_name': 'device type',
                'verbose_name_plural': 'device types',
            },
        ),
        migrations.CreateModel(
            name='SubAccount',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('account_number', models.CharField(help_text='\u5b50\u8d26\u6237\u7f16\u53f7', unique=True, max_length=169)),
                ('user_id', models.CharField(help_text='\u7528\u6237id', max_length=128)),
                ('time_created', models.DateTimeField(help_text='\u8d26\u6237\u521b\u5efa\u65f6\u95f4', auto_now_add=True)),
                ('currency', models.CharField(default=b'CNY', help_text='\u4ea4\u6613\u5e01\u79cd', max_length=8, choices=[(b'CNY', b'\xe4\xba\xba\xe6\xb0\x91\xe5\xb8\x81'), (b'JCC', b'\xe8\x8a\x82\xe6\x93\x8d\xe5\xb8\x81')])),
                ('balance', models.DecimalField(default=Decimal('0.000000'), help_text='\u8d26\u6237\u4f59\u989d', null=True, max_digits=12, decimal_places=6)),
                ('status', models.CharField(default=b'OPEN', help_text='\u8d26\u6237\u72b6\u6001', max_length=32, choices=[(b'OPEN', b'\xe5\xbc\x80\xe5\x90\xaf'), (b'FROZEN', b'\xe5\x86\xbb\xe7\xbb\x93'), (b'CLOSED', b'\xe5\x85\xb3\xe9\x97\xad')])),
                ('account', models.ForeignKey(related_name='sub_accounts', to='accounts.Account')),
            ],
            options={
                'verbose_name': 'sub account',
                'verbose_name_plural': 'sub accounts',
            },
        ),
        migrations.CreateModel(
            name='SystemCode',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('sys_code_en', models.CharField(help_text='\u82f1\u6587\u7cfb\u7edf\u4ee3\u7801', max_length=128)),
                ('sys_code_cn', models.CharField(help_text='\u4e2d\u6587\u7cfb\u7edf\u4ee3\u7801', max_length=128)),
                ('mapping_code', models.CharField(help_text='\u6620\u5c04\u7f16\u53f7', unique=True, max_length=32)),
            ],
            options={
                'verbose_name': 'system code',
                'verbose_name_plural': 'system codes',
            },
        ),
        migrations.CreateModel(
            name='Transaction',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('amount', models.DecimalField(max_digits=12, decimal_places=6)),
                ('time_created', models.DateTimeField(auto_now_add=True)),
                ('sub_account', models.ForeignKey(related_name='transactions', to='accounts.SubAccount')),
            ],
            options={
                'verbose_name': 'transaction',
                'verbose_name_plural': 'transactions',
            },
        ),
        migrations.CreateModel(
            name='Transfer',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('source_balance', models.DecimalField(default=Decimal('0.000000'), help_text='\u8f6c\u8d26\u540e\u7684\u6765\u6e90\u65b9\u4f59\u989d', max_digits=12, decimal_places=6)),
                ('destination_balance', models.DecimalField(default=Decimal('0.000000'), help_text='\u8f6c\u8d26\u540e\u7684\u63a5\u53d7\u65b9\u4f59\u989d', max_digits=12, decimal_places=6)),
                ('amount', models.DecimalField(max_digits=12, decimal_places=6)),
                ('description', models.CharField(max_length=256, null=True)),
                ('time_created', models.DateTimeField(auto_now_add=True)),
                ('object_id', models.PositiveIntegerField()),
                ('content_type', models.ForeignKey(to='contenttypes.ContentType')),
                ('destination', models.ForeignKey(related_name='destination_transfers', to='accounts.SubAccount')),
                ('source', models.ForeignKey(related_name='source_transfers', to='accounts.SubAccount')),
            ],
            options={
                'verbose_name': 'transfer',
                'verbose_name_plural': 'transfers',
            },
        ),
        migrations.AddField(
            model_name='transaction',
            name='transfer',
            field=models.ForeignKey(related_name='transactions', to='accounts.Transfer'),
        ),
        migrations.AlterUniqueTogether(
            name='transaction',
            unique_together=set([('transfer', 'sub_account')]),
        ),
    ]
